src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,199 @@
+        /// </remarks>
+        public int IndexOf([NotNull] TAttribute item)
+        {
+            return Attributes.IndexOf(item);
+        }
+
+        /// <summary>
+        /// Retrieves the first <typeparamref name="TAttribute"/> whose <typeparamref name="TAttribute.Name"/> matches
+        /// the specified
+        /// <paramref name="name"/>.
+        /// </summary>
+        /// <param name="name">The <typeparamref name="TAttribute.Name"/> of the <typeparamref name="TAttribute"/> to
+        /// get.</param>
+        /// <param name="attribute">When this method returns, the first <typeparamref name="TAttribute"/> whose
+        /// <typeparamref name="TAttribute.Name"/> matches the specified <paramref name="name"/>, if found; otherwise,
+        /// <c>null</c>.</param>
+        /// <returns><c>true</c> if a <typeparamref name="TAttribute"/> with the same
+        /// <typeparamref name="TAttribute.Name"/> exists in the collection; otherwise, <c>false</c>.</returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively.</remarks>
+        public bool TryGetAttribute([NotNull] string name, out TAttribute attribute)
+        {
+            attribute = Attributes.FirstOrDefault(attr => NameEquals(name, attr));
+
+            return attribute != null;
+        }
+
+        /// <summary>
+        /// Retrieves <typeparamref name="TAttribute"/>s in the collection whose <typeparamref name="TAttribute.Name"/>
+        /// matches the specified <paramref name="name"/>.
+        /// </summary>
+        /// <param name="name">The <typeparamref name="TAttribute.Name"/> of the <typeparamref name="TAttribute"/>s to
+        /// get.</param>
+        /// <param name="attributes">When this method returns, the <typeparamref name="TAttribute"/>s whose
+        /// <typeparamref name="TAttribute.Name"/> match the specified <paramref name="name"/>, if at least one is
+        /// found; otherwise, <c>null</c>.</param>
+        /// <returns><c>true</c> if at least one <typeparamref name="TAttribute"/> with the same
+        /// <typeparamref name="TAttribute.Name"/> exists in the collection; otherwise, <c>false</c>.</returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively.</remarks>
+        public bool TryGetAttributes([NotNull] string name, out IEnumerable<TAttribute> attributes)
+        {
+            attributes = Attributes.Where(attribute => NameEquals(name, attribute));
+
+            return attributes.Any();
+        }
+
+        /// <inheritdoc />
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        /// <inheritdoc />
+        public IEnumerator<TAttribute> GetEnumerator()
+        {
+            return Attributes.GetEnumerator();
+        }
+
+        internal static bool NameEquals(string name, [NotNull] TAttribute attribute)
```

Comment:
we want to expose it to subclasses and we don't need it for testing -- the only reason we use `internal` anywhere.